8주차 레스토랑 앱 꾸미기

solmatches opened the pull request:
안녕하세요! 마지막 과제 제출을 시작합니다ㅠ

Redux Toolkit의 createSlice 사용 및 duck pattern을 적용하였고
createAsyncThunk을 이용해 비동기 작업을 리팩토링 해보았습니다. 

내일은 createAsyncThunk 적용으로 인해 변경된 비동기 액션 테스트를 어떻게 업데이트 해야할지 고민해보고 적용해볼 예정입니다!

solmatches committed:
Install emotion/styled, emoition/react

solmatches committed:
Add reduxjs/toolkit

solmatches committed:
Change redux to redux toolkit

solmatches committed:
Update slice.js to duck patterns

solmatches committed:
Update Async Requests width createAsyncThunk

hannut91 commented on src/slice.js:
```diff
@@ -0,0 +1,254 @@
+  },
+  extraReducers: (builder) => {
+    builder
+      .addCase(loadInitialData.fulfilled, (state, { payload: { regions, categories } }) => {
+        state.regions = regions;
+        state.categories = categories;
+      })
```
빌더를 사용하는 것보다 다음과 같이 쓰는게 더 편하지 않나용?

```js
const reducer1 = createReducer(initialState, {
  extraReducers: {
    [loadInitialDatafulfilled]: (state, action) => {
      // ... 생략
    },
  }
})
```

See also
  - https://redux-toolkit.js.org/api/createAsyncThunk

solmatches replied:
오옷 확실히 더 편하고 보기도 좋네요! 코드숨에서 배웠던 내용과도 일맥상통하고요
객체키 로 표현할 수 있는 표기법이 있었는데 미처 발견하지 못했었네요 알려주셔서 감사합니다! 👍👍

solmatches committed:
Change extraReducers to object key notation

solmatches committed:
Update slice and slice.test

solmatches commented:
오늘은 createAsyncThunk가 적용된 slice 테스트 작성 방법에 대해서 종일 알아보다가 시간을 다 보내버렸네요 🥲

hannut91 commented on src/slice.js:
```diff
@@ -0,0 +1,198 @@
+  extraReducers: {
+    [loadInitialData.fulfilled]: (state, { payload: { regions, categories } }) => {
+      state.regions = regions;
+      state.categories = categories;
```
redux에서 immerjs를 사용해서 mutable하게 코드를 작성하면서 immutable하게 동작하도록 하는 코드인 것 같아요. 문법적으로 편리하지만, 안좋은 습관을 만들 수 있어서 저는 강력하게 반대하는 편입니다.

solmatches replied:
문서를 더 찾아보니 지금까지 했던 것 처럼 state를 복사해서 새로운 state를 return 하는 방법으로도 사용할 수 있더라구요! 안 그래도 reassign으로 인해 lint 오류가 났었는데 disabled 시키면서도 조금 걸렸었는데.. 안 좋은 습관을 만들 수 있다는 말씀에 경각심이 생겨서 더 찾아보고 좋은 방법을 찾을 수 있었습니다!

hannut91 commented on src/slice.test.js:
```diff
@@ -0,0 +1,335 @@
+      });
+
+      it('did not change restaurants', async () => {
+        const action = await store.dispatch(loadRestaurants());
```
async액션이 액션을 반환하고 액션을 실행했을 때를 테스트하셨군요. 이 액션이 무엇이 나오는지만 테스트하고, 해당 액션은 따로 테스트해도 좋았을 것 같아요.

hannut91 reviewed (commented):
`React.memo()`를 올바르게 사용하는 방법은 명확하게 아셔야 합니다. 잘못사용하면 올바르게 동작하지 않아서 사용하지 않는 것보다 더 안 좋아질 수 있어요. 예를들어서 `React.memo`를 사용했는데 props로 익명함수를 받아서 매번 값이 달라진다고 판단하여 다시 렌더링하고 다시 메모이제이션을 하면 오히려 안하는 것보다 낫겠죠. 이렇게 제대로 사용하는 방법은 완전히 익히셔야 합니다.

근데 이 `React.memo`를 사용해야 하는 때는 명확한 기준은 없습니다. 그냥 명목적으로 적용하면 구현하는데 시간만 더 걸리고 복잡해지기만 합니다. 그래서 `측정`이 중요합니다. 성능을 측정하고, 성능을 얼마나 개선할지 목표를 정하고 실제로 개선한 후 다시 측정해서 실제로 효과가 있는지를 확인해야 하는게 정석적인 방법입니다.

또 생각해볼점은 처음부터 성능을 고려하는 것은 별로 좋지 않습니다. 왜냐하면 너무 이른 최적화는 구현을 복잡하게 하기 때문입니다. 아주 필요한 경우에만 하는게 좋습니다. 지금은 연습하는 과정이니 사용하는 법을 충분히 익혀두고, 나중에 필요할 때 적절히 사용해야 합니다. 어떻게 측정해야 하는지도 공부해 보세요. 

https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en

solmatches committed:
Refactor extraReducer return state

mutable하게 코드를 작성하는 방법은 편리하지만 안 좋은 습관이 생길 수 있다는 리뷰에 공감하여
새로운 state를 return 하는 방법으로 변경

solmatches committed:
Change directory hierarchy

solmatches committed:
Add Header component test

solmatches committed:
Update react-testing-library, react-router package

solmatches committed:
Add Header component, GlobalStyle

solmatches committed:
Add Button component test

solmatches committed:
Add Button component

solmatches committed:
Update Components Styles

solmatches committed:
Update useCallback to eventHandlers

solmatches committed:
Update package-lock.json, yarn.lock

github action 실행시 npm 으로 빌드됨.
로컬에서는 yarn으로 사용하다보니
package.json에 있는 내용과 package-lock.json 내용이 달라 액션이 실패함
